Lakat/src/Hooks.php

```diff
@@ -19,6 +19,7 @@
 
 namespace MediaWiki\Extension\Lakat;
 
+use MediaWiki\Title\Title;
```

Comment:
General question about php. This brings is some namespace stuff that brings the Title object into scope, right? How do i read this path? Is Title a submodule of MediaWiki and Title an Object in the module Title?

Reply:
> This brings is some namespace stuff that brings the Title object into scope, right?

Right. This imports class `\MediaWiki\Title\Title` (fully qualified name of the class with full namespace prefix) into Hooks.php and gives it a short alias `Title` that can be used in code. Otherwise, without `use` statement you have to reference fully qualified class name in code.

> How do i read this path?

TLDR: classes from `\MediaWiki\` namespace can be found in `/includes/` dir. Root path of namespace is not necessarely the same as path in filesystem.

This thing is closely related to autoloading of classes. If autoloading is not used, then you have to directly include one php file into another. PHP uses autoloader function to calculate class file path based on it's namespace. PHP has default [autoloader](https://www.php.net/manual/ru/function.spl-autoload.php) that searches for classes in current dir and configured library include paths. Mediawiki, like many other projects, replaces PHP default autoloader with [its own](https://www.mediawiki.org/wiki/Manual:AutoLoader.php). In that file you can see mapping of namespaces to file paths for MediaWiki files:
```
'MediaWiki\\' => __DIR__ . '/',
```
where __DIR__ is `/includes/` dir in mediawiki installation.
You can also find mapping for each class individually in automatically generated autoload.php file:
```
'MediaWiki\\Title\\Title' => __DIR__ . '/includes/title/Title.php',
```

For external libraries that are installed with composer, composer creates it's own autoloader, all those files can be found in `/vendor/` dir in relevant library subdir.

> Is Title a submodule of MediaWiki and Title an Object in the module Title?

Kind of, but I wouldn't call it modules because PHP doesn't have real modules contrary to Python. In PHP world they're namespaces and classes in those namespaces. Namespaces are used to arrange classes just like directory hierarchy is used to arrange files.

You correctly assumed that's the common agreement is to use the name of the project as the first part of the namespace. Following parts are usually subdirectories in that project and reflect some logical arrangement of classes.